src/ripple/conditions/impl/fulfillment.cpp

```diff
@@ -0,0 +1,211 @@
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/conditions/condition.h>
+#include <ripple/conditions/fulfillment.h>
+#include <ripple/conditions/hashlock.h>
+#include <ripple/conditions/prefix_sha256.h>
+#include <ripple/conditions/ed25519.h>
+#include <ripple/conditions/rsa_sha256.h>
+#include <ripple/conditions/impl/utils.h>
+#include <boost/regex.hpp>
+#include <boost/optional.hpp>
+#include <type_traits>
+#include <vector>
+
+namespace ripple {
+namespace cryptoconditions {
+
+bool
+fulfills (
+    fulfillment_t const& f,
+    condition_t const& c)
+{
+    // Fast check: the fulfillment's type must match the
+    // conditions's type:
+    if (f.type() != c.type)
+        return false;
+
+    return f.condition() == c;
+}
+
+bool
+validate (
+    fulfillment_t const& f,
+    condition_t const& c,
+    Slice const& m)
+{
+    if (!fulfills (f, c))
+        return false;
+
+    // Check if the message validates:
+    return f.validate (m);
+}
+
+bool
+validate (
+    fulfillment_t const& f,
+    Slice const& m)
+{
+    return validate (f, f.condition(), m);
+}
+
+std::unique_ptr<fulfillment_t>
+load_fulfillment (std::uint16_t type, Slice payload)
+{
+    std::unique_ptr<fulfillment_t> p;
+
+    switch (type)
+    {
+    case condition_hashlock:
+        p = std::make_unique<hashlock_t>();
+        break;
+
+    case condition_prefix_sha256:
+        p = std::make_unique<prefix_sha256_t>();
+        break;
+
+    case condition_rsa_sha256:
+        p = std::make_unique<rsa_sha256_t>();
+        break;
+
+    case condition_ed25519:
+        p = std::make_unique<ed25519_t>();
+        break;
+
+    default:
+        throw std::domain_error (
+            "Unknown cryptocondition type " +
+                std::to_string (type));
+    }
+
+    // If the payload can't be parsed, the load should
+    // fail.
+    if (p && !p->parse_payload(payload))
+        p.reset();
+
+    return p;
+}
+
+// Parse a condition from its string form
+std::unique_ptr<fulfillment_t>
+load_fulfillment (std::string const& s)
+{
+    // CHECKME: custom parser maybe? probably faster but
+    // more work and probability of error.
```

Comment:
Not for this changeset, but we should see what a custom parser looks like for this. The fields are delimited by ':' so a strategy of finding the delimiters and parsing the fields should work well (and would copy fewer strings). We could probably reuse the parser for conditions as well.